My first commit

Porestoranam opened the pull request:
first attemp

Commit:
My first commit

WokinShopash commented on encode_and_decode.py:
```diff
@@ -0,0 +1,53 @@
+def new_char(char, shift):      # make new char using value of shift
```
Лучше такие комментарии писать как docstring, на строчке сразу после объявления

Porestoranam replied:
Все комментарии перенесены в docstring для каждой функции.

WokinShopash commented on encryptor.py:
```diff
@@ -0,0 +1,100 @@
+        input_ = input()
+    if args.cipher == 'caesar':
+        if args.output_file is not None:
+            output = open(args.output_file, 'w')
```
Избавься от дублирования кода, отвечающего за чтение/запись

Porestoranam replied:
Сделано. Теперь код, отвечающий за чтение и запись находится только в одном месте.

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,86 @@
+import string
+
+
+def new_char(char, shift):      # make new char with using value of shift
```
Можно эту функцию тоже заимпортить из соседнего файла. Но дублировать точно не надо

Porestoranam replied:
Заимпорчено:)

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,86 @@
+    dic = dict()
+    for symbol in string.ascii_letters:
+        dic[symbol] = 0
+    dic['length'] = 0
```
Разве значение этого ключа ключа где-то потом используется?

Porestoranam replied:
Действительно, не используется, спасибо!:)

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,86 @@
+    for symbol in main_string:
+        if symbol in string.ascii_letters:
+            allocation_for_string[symbol] += 1
+    length = len(main_string)
```
Если такой знаменатель брать, то сумма столбцов будет не 1

Porestoranam replied:
Да, были учтены абсолютно все символы, даже те, которые не являются буквами. Исправлено.

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def answer(main_string, ideal_allocation):  # return ans_string
+    ans = 106.0
```
Какая-то магическая константа

Porestoranam replied:
Это просто максимальная граница, которая возможна. Всего различных букв получается 52, отклонение каждой из которых по модулю не больше 2. Поставил sys.maxsize

Commit:
change docstring and attemp to change double output and input files

Commit:
changed double code of reading and writing

Commit:
with ideal alloc

Commit:
fixed some unuseful objects
changed values of allocation (sum of coloumns is right( = 1) now)

Commit:
magic const is changed into sys.maxsize
we cann't have value more than 106 because there are 52 different digits and max deviation is 2.(52 * 2 < 106)
also it is last commit. All demands are executed

Commit:
updated with counter and str.join()

Commit:
updated with counter and str.join()

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def get_all_opportunities(main_string):
+    """get all strings we can"""
```
Ууу, даже если забыть про то, что мы на паре выяснили, что проходиться можно всего 2 раза вместо 27, 
хранить все варианты раскодирования - ну ооочень затратно

Porestoranam replied:
Исправлено. Да, спасибо, действительно, алгоритмически лучше, да и кода получилось меньше)

WokinShopash commented on encode_and_decode.py:
```diff
@@ -0,0 +1,46 @@
+    dis_from_a = ord_char - ord('a')
+    dis_from_z = ord('z') - ord_char
+    if (dis_from_cap_a >= 0) and (dis_from_cap_z >= 0):
+        new_ord = (ord_char - ord('A') + shift) % 26 + ord('A')
```
26 - магическая константа, которая должна быть унесена куда-то наверх, объявлена один раз с хорошим именем и использована везде в других местах
А еще лучше вычислить ее с помощью того же модуля string

Porestoranam replied:
Сделано

WokinShopash commented on encode_and_decode.py:
```diff
@@ -0,0 +1,46 @@
+def new_char(char, shift):
+    """make new char using value of shift"""
+    ord_char = ord(char)
+    dis_from_cap_a = ord_char - ord('A')
```
Какой смысл загонять эти выражения в отдельные переменные, если потом используешь их только по одному разу через 3 строчки в условии, не особо уменьшив количество символов. 
Учитывая, что св питоне можно писать конструкции вида a < b < c, так это и вовсе становится ненужным

Porestoranam replied:
И в правду, только лишний код:)

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,64 @@
+
+def count_approximate(main_string):
+    """return dict of allocation(v) letters for main_string"""
+    allocation_for_string = collections.Counter(ch for ch in main_string if ch in string.ascii_letters)
```
Пожалуй, строчные и заглавные буквы мы тут хотим считать в одной куче. заглавных в обычном слишком мало, чтобы они дали какую-то хорошую статистику

Porestoranam replied:
Теперь все буквы считаются маленькими и по ним идет апроксимация

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,64 @@
+    return opportunities_strings
+
+
+def count_ideal_allocation(text_for_train):
```
Эта функция абсолютно такая же, как count_approximate
Дублирование кода, печалька

Porestoranam replied:
Лучше было вообще убрать эти функции, поскольку для реализации алгоритмически выгодного варианта, они только мешают)

WokinShopash replied:
Как-то ты не убрал их в итоге.
Смотри, логика того, как считается частота, у тебя сейчас встречается дважды. Надо, чтобы было один раз.

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,64 @@
+    return ideal_allocation
+
+
+def make_dict_of_strings_and_values(main_string, ideal_allocation):
```
Делать большие строчки ключами словаря - плохая идея.
Поиск по такому словарю сначала проверяет равенство хешей, а потом посимвольное равенство строк, которое займет еще O(длина_текста) времени.

Да и вообще, зачем считать заранее все разницы с эталоном для всех вариантов сдвига? На лету же можно минимум обновлять. 
Если тебе хочется разделить генерацию текста, подсчет разницы с эталоном на разные функции, напиши генераторы.

Porestoranam replied:
Убрал)

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,64 @@
+
+def answer(main_string, ideal_allocation):
+    """return ans_string"""
+    ans = sys.maxsize
```
Лучше даже float("inf") - бесконечность в самом что ни на есть классическом понимании

Porestoranam replied:
Дас)

Commit:
made short code

Commit:
made short code

WokinShopash commented on hack_and_train.py:
```diff
@@ -0,0 +1,46 @@
+def changed_text(main_string, ideal_allocation):
+    """returns changed_text with the best approximation"""
+    ans_shift = answer_shift_func(main_string, ideal_allocation)
+    ans_text = []
```
А с это места здесь с точностью до названий переменных скопирован код ф-ции caesar_encode_string

Porestoranam replied:
Точно)

WokinShopash commented on encryptor.py:
```diff
@@ -0,0 +1,82 @@
+
+
+elif args.method == 'train':
+    if args.text_file is not None:
```
Сейчас эта логика точь-в-точь повторяет логику чтения на строке 37.
Надо бы сделать это отдельной функцией.

Porestoranam replied:
Сделано)

WokinShopash commented on encryptor.py:
```diff
@@ -0,0 +1,82 @@
+
+if hasattr(args, 'output_file'):
+    if args.output_file is not None:
+        output = open(args.output_file, 'w')
```
Вся запись в файл может делаться только под with

Porestoranam replied:
fixed

WokinShopash commented on encryptor.py:
```diff
@@ -0,0 +1,82 @@
+    else:
+        text_for_train = input()
+    allocation_in_model = hack_and_train.count_ideal_allocation(text_for_train)
+    pickle.dump(allocation_in_model, open(args.model_file, 'wb'))
```
Вся запись в файл может делаться только под with

Porestoranam replied:
fixed

WokinShopash commented on encryptor.py:
```diff
@@ -0,0 +1,82 @@
+    else:
+        input_ = input()
+
+
```
Надо избавиться от кода в глобальном namespace. Весь логический код должен быть разделен на функции
Создать функцию для определения и считывания аргументов (get_args), и функция для выполнения логики (main).
Потом последовательно вызвать с помощью `__name__`-проверки, как вот тут:
https://github.com/WokinShopash/FlaskDemoLazyStudents/blob/master/main.py#L93

Porestoranam replied:
Выделил функцию, теперь мэйн выполняет логику

Commit:
fixed last code mistakes

WokinShopash reviewed (approved):
Зачтено.
